File: backend/agents/opportunity.py
"""
Opportunity Agent
Converts high-frequency complaints into actionable product opportunities.
Each pattern maps a pain signal to a specific product feature recommendation.
"""
import time
import re
from memory import SharedMemory, AgentLog

OPP_PATTERNS = [
    (r"refund|return|money back", "Build real-time refund status tracker with automated SLA notifications"),
    (r"crash|bug|error|fail|broken", "Implement crash analytics and proactive user error recovery flows"),
    (r"support|response|help|chat|agent", "Deploy AI-powered instant support with < 30s first response guarantee"),
    (r"payment|transaction|fail|upi", "Build payment retry intelligence with smart gateway fallback routing"),
    (r"delivery|late|track|status", "Create live delivery tracking with proactive delay alerts and compensation"),
    (r"price|discount|expensive|cheaper|cost", "Introduce transparent pricing dashboard with loyalty reward visibility"),
    (r"account|login|otp|blocked", "Implement frictionless re-authentication with biometric fallback options"),
    (r"slow|performance|loading|fast", "Invest in performance optimization targeting < 2s load time globally"),
    (r"onboard|setup|confus|difficult|first", "Build contextual onboarding wizard with role-based personalization"),
    (r"data|export|download|report", "Add self-serve analytics dashboard with one-click data export (CSV/JSON)"),
]


def run_opportunity(memory: SharedMemory) -> None:
    start = time.time()
    memory.agent_logs.append(AgentLog("Opportunity Agent", "running", "Mapping pain signals to product opportunities...", time.time()))

    opp_count = 0
    for signal in memory.signals:
        opportunities = []
        search_text = f"{signal.label} {' '.join(signal.risks)} {signal.description}".lower()

        for pattern, opp in OPP_PATTERNS:
            if re.search(pattern, search_text):
                opportunities.append(opp)

        # High volume + high negativity = validated unmet need
        if signal.negativity >= 60 and signal.volume >= 3:
            opportunities.append(
                f"High complaint frequency ({signal.volume} mentions, {signal.negativity}% negativity) "
                "validates a clear market gap — prioritize for roadmap Q1"
            )

        signal.opportunities = list(dict.fromkeys(opportunities))
        if signal.opportunities:
            signal.is_opportunity = True
            opp_count += 1

    duration = int((time.time() - start) * 1000)
    memory.agent_logs.append(AgentLog(
        "Opportunity Agent", "done",
        f"Found {opp_count} opportunity signals from {len(memory.signals)} clusters",
        time.time(), duration
    ))

File: Insight/src/agents/opportunity.ts
/**
 * Opportunity Agent (Tier 2 Evaluator)
 * Identifies unmet needs and repeated feature requests from pain signals.
 */
import type { Signal } from '@/types';

const OPPORTUNITY_PATTERNS = [
  { pattern: /support|response|ticket/i, opp: 'Build AI-powered instant support with knowledge base integration' },
  { pattern: /onboard|setup|learn|confus/i, opp: 'Interactive onboarding wizard with personalized user journey' },
  { pattern: /price|pricing|cost|transparent/i, opp: 'Self-serve pricing calculator with transparent tier comparison' },
  { pattern: /integrat|api|connect|zapier/i, opp: 'Native integration marketplace with no-code connectors' },
  { pattern: /mobile|app|phone/i, opp: 'Mobile-first product redesign with feature parity' },
  { pattern: /data|export|portab/i, opp: 'One-click data export in multiple formats as standard feature' },
  { pattern: /ai|artificial|search/i, opp: 'Reliable AI features with clear capability boundaries and user trust' },
  { pattern: /trial|demo|evaluat/i, opp: 'Frictionless 30-day trial with no credit card required' },
  { pattern: /email|notif|overwhelm/i, opp: 'Smart notification system with user-controlled frequency preferences' },
];

export async function runOpportunityAgent(
  signals: Signal[],
  onProgress: (msg: string) => void,
): Promise<Signal[]> {
  onProgress('Opportunity Agent: Identifying unmet needs and feature gaps...');
  await new Promise(r => setTimeout(r, 150));

  const enriched = signals.map(signal => {
    const opportunities: string[] = [];
    const searchText = `${signal.label} ${signal.description} ${signal.posts.slice(0, 5).map(p => p.title).join(' ')}`;

    for (const { pattern, opp } of OPPORTUNITY_PATTERNS) {
      if (pattern.test(searchText)) opportunities.push(opp);
    }

    // High negativity + high volume = validated market gap
    if (signal.negativity >= 70 && signal.volume >= 3) {
      opportunities.push('High pain × high frequency = validated unmet market need ready for product investment');
    }

    return {
      ...signal,
      opportunities: [...new Set(opportunities)],
      isOpportunity: opportunities.length > 0,
    };
  });

  const oppCount = enriched.filter(s => s.isOpportunity).length;
  onProgress(`Opportunity Agent: Found ${oppCount} opportunity signals ✓`);
  return enriched;
}
